transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java

```diff
@@ -1258,55 +1238,14 @@ public boolean spliceIn(RecvByteBufAllocator.Handle handle) {
         }
     }
 
-    private final class SocketWritableByteChannel implements WritableByteChannel {
-
-        @Override
-        public int write(ByteBuffer src) throws IOException {
-            final int written;
-            int position = src.position();
-            int limit = src.limit();
-            if (src.isDirect()) {
-                written = fd().write(src, position, src.limit());
-            } else {
-                final int readableBytes = limit - position;
-                ByteBuf buffer = null;
-                try {
-                    if (readableBytes == 0) {
-                        buffer = Unpooled.EMPTY_BUFFER;
-                    } else {
-                        final ByteBufAllocator alloc = alloc();
-                        if (alloc.isDirectBufferPooled()) {
-                            buffer = alloc.directBuffer(readableBytes);
-                        } else {
-                            buffer = ByteBufUtil.threadLocalDirectBuffer();
-                            if (buffer == null) {
-                                buffer = Unpooled.directBuffer(readableBytes);
-                            }
-                        }
-                    }
-                    buffer.writeBytes(src.duplicate());
-                    ByteBuffer nioBuffer = buffer.internalNioBuffer(buffer.readerIndex(), readableBytes);
-                    written = fd().write(nioBuffer, nioBuffer.position(), nioBuffer.limit());
-                } finally {
-                    if (buffer != null) {
-                        buffer.release();
-                    }
-                }
-            }
-            if (written > 0) {
-                src.position(position + written);
-            }
-            return written;
-        }
-
-        @Override
-        public boolean isOpen() {
-            return fd().isOpen();
+    private final class EpollSocketWritableByteChannel extends SocketWritableByteChannel {
+        EpollSocketWritableByteChannel() {
+            super(socket);
         }
 
         @Override
-        public void close() throws IOException {
-            fd().close();
+        protected ByteBufAllocator alloc() {
+            return alloc();
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) There is an apparent infinite recursive loop in io.netty.channel.epoll.AbstractEpollStreamChannel$EpollSocketWritableByteChannel.alloc() [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=findbugs%3AIL_INFINITE_RECURSIVE_LOOP)